### PR TITLE
On E0204 suggest missing type param bounds

### LINKED ISSUE #89137: Show error E0277 when the trait bound of a struct's member's type is unsatisfied even if E0204 is also present
Given the following code: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=fbd47fd0a07c3a9a05978696f46aac3d

```rust
use std::fmt::Debug;

#[derive(Debug, Copy, Clone)]
pub struct Vector2<T: Debug + Copy + Clone>{
    pub x: T,
    pub y: T
}

#[derive(Debug, Copy, Clone)]
pub struct AABB<T>{
    pub loc: Vector2<T>,
    pub size: Vector2<T>
}
```

The current output is:

```
error[E0204]: the trait `Copy` may not be implemented for this type
  --> src/main.rs:10:17
   |
10 | #[derive(Debug, Copy, Clone)]
   |                 ^^^^
11 | pub struct AABB<T>{
12 |     pub loc: Vector2<T>,
   |     ------------------- this field does not implement `Copy`
13 |     pub size: Vector2<T>
   |     -------------------- this field does not implement `Copy`
   |
   = note: this error originates in the derive macro `Copy` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0204`.
```

Ideally the output should look like:

Keeping the other error may or may not make sense, but I propose adding this, as it points more towards how to fix the issue and prevents barking up the tree of how a derive(Copy)'d type can not implement copy (there's a good answer here, but it was not immediately obvious to me upon seeing the original error).

I generated this error message by removing the derive line from the AABB struct.

```
error[E0277]: the trait bound `T: Copy` is not satisfied
  --> src/lib.rs:10:14
   |
4  | pub struct Vector2<T: Debug + Copy + Clone>{
   |                               ---- required by this bound in `Vector2`
...
10 |     pub loc: Vector2<T>,
   |              ^^^^^^^^^^ the trait `Copy` is not implemented for `T`
   |
help: consider restricting type parameter `T`
   |
9  | pub struct AABB<T: std::marker::Copy>{
   |                  ^^^^^^^^^^^^^^^^^^^

For more information about this error, try `rustc --explain E0277`.
error: could not compile `playground` due to 2 previous errors
```

This bug takes place on beta and nightly according to the Rust playground.

### PR DESCRIPTION
**Pull request by estebank**
```
error[E0204]: the trait `Copy` may not be implemented for this type
  --> f42.rs:9:17
   |
9  | #[derive(Debug, Copy, Clone)]
   |                 ^^^^
10 | pub struct AABB<K>{
11 |     pub loc: Vector2<K>,
   |     ------------------- this field does not implement `Copy`
12 |     pub size: Vector2<K>
   |     -------------------- this field does not implement `Copy`
   |
note: the `Copy` impl for `Vector2<K>` requires that `K: Debug`
  --> f42.rs:11:5
   |
11 |     pub loc: Vector2<K>,
   |     ^^^^^^^^^^^^^^^^^^^
note: the `Copy` impl for `Vector2<K>` requires that `K: Debug`
  --> f42.rs:12:5
   |
12 |     pub size: Vector2<K>
   |     ^^^^^^^^^^^^^^^^^^^^
   = note: this error originates in the derive macro `Copy` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider restricting type parameter `K`
   |
10 | pub struct AABB<K: Debug>{
   |                  +++++++
```

Fix #89137.

### CODE REVIEW

**Comment by compiler-errors** [comment]
This is very cool

@bors r+ rollup

**Comment by compiler-errors** [comment]
@bors r-

r=me with the test fixed :smiley_cat:

**Comment by estebank** [comment]
@compiler-errors added another commit with a small, hopefully uncontroversial change.

@bors r=compiler-errors

**Comment by compiler-errors** [comment]
Looks good to me, other than if the usage of `FxHashMap` there is going to cause hash instability in diagnostics. Maybe we should stringify first?

**Comment by compiler-errors** [comment]
oh... `stable_map`? Did you mean to use this one? https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_data_structures/stable_map.rs.html#7

**Comment by estebank** [comment]
Changed :)

@bors r=compiler-errors

**Comment by estebank** [comment]
🤦 I'd failed to push

**Comment by compiler-errors** [comment]
There we go. Thanks :+1:

**Comment by compiler-errors** [comment]
Oh no, haha. @bors r-

I think it's best to just `.to_string()` on the two keys for the map and use something sorted like a `BTreeMap`?

**Comment by estebank** [comment]
I'll make it a string in a bit

File: compiler/rustc_typeck/src/coherence/builtin.rs [hunk 1]
```diff
@@ -115,17 +129,46 @@ fn visit_implementation_of_copy(tcx: TyCtxt<'_>, impl_did: LocalDefId) {
...
                         if error_predicate != error.root_obligation.predicate {
-                            err.span_note(
-                                error.obligation.cause.span,
-                                &format!(
-                                    "the `Copy` impl for `{}` requires that `{}`",
-                                    ty, error_predicate
-                                ),
-                            );
+                            errors
+                                .entry((ty, error_predicate))
```
Comment (compiler-errors) [inline compiler/rustc_typeck/src/coherence/builtin.rs:133]:
```suggestion
                            errors
                                .entry((ty.to_string(), error_predicate.to_string()))
```

File: compiler/rustc_typeck/src/coherence/builtin.rs [hunk 2]
```diff
@@ -91,6 +91,20 @@ fn visit_implementation_of_copy(tcx: TyCtxt<'_>, impl_did: LocalDefId) {
...
+            // their `Copy` implementation.
+            let mut generics = None;
+            if let ty::Adt(def, _substs) = self_type.kind() {
+                let self_def_id = def.did();
+                if let Some(local) = self_def_id.as_local() {
+                    let self_item = tcx.hir().expect_item(local);
+                    generics = self_item.kind.generics();
+                }
+            }
+            let mut errors: StableMap<_, Vec<_>> = Default::default();
```
Comment (compiler-errors) [inline compiler/rustc_typeck/src/coherence/builtin.rs:105]:
```suggestion
            let mut errors: BTreeMap<_, Vec<_>> = Default::default();
```

**Review by compiler-errors — COMMENTED** [review]
Something like:

**Comment by compiler-errors** [comment]
CI seems happy!

@bors r+

**Pull request merged**